Gem/Code/Source/Components/UI/MatchOverComponent.cpp

```diff
@@ -0,0 +1,118 @@
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::Category, "Multiplayer Sample UI")
+                    ->Attribute(AZ::Edit::Attributes::Icon, "Icons/Components/Component_Placeholder.svg")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC_CE("Game"))
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default, &MatchOverComponent::m_gameOverElementId, "Game Over element ID", "The element ID of the Game Over UI element to display at Match end.")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default, &MatchOverComponent::m_hudElementId, "HUD element ID", "The element ID of the UI element for the HUD")
+                    ;
+            }
+        }
+    }
+
+    void MatchOverComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC("NetworkMatchComponent"));
+        required.push_back(AZ_CRC("UiCanvasRefService"));
+    }
+
+    void MatchOverComponent::Activate()
+    {
+        UiCanvasRefBus::EventResult(m_uiCanvasId, GetEntityId(), &UiCanvasRefBus::Events::GetCanvas);
+        if (!m_uiCanvasId.IsValid())
+        {
+            UiCanvasAssetRefNotificationBus::Handler::BusConnect(GetEntityId());
+        }
+
+        if (NetworkMatchComponent* netMatchComponent = GetEntity()->FindComponent<NetworkMatchComponent>())
+        {
+            m_roundNumberHandler = AZ::EventHandler<uint16_t>([this](uint16_t value) 
+                { UpdateRound(value); });
+            netMatchComponent->RoundNumberAddEvent(m_roundNumberHandler);
+
+            m_roundTimerHandler = AZ::EventHandler<RoundTimeSec>([this](RoundTimeSec value) 
+                { DetermineIfMatchEnded(value); });
+            netMatchComponent->RoundTimeAddEvent(m_roundTimerHandler);
+        }
+    }
+
+    void MatchOverComponent::Deactivate()
+    {
+        UiCanvasAssetRefNotificationBus::Handler::BusDisconnect(GetEntityId());
+    }
+
+    void MatchOverComponent::OnCanvasLoadedIntoEntity(AZ::EntityId uiCanvasEntity)
+    {
+        m_uiCanvasId = uiCanvasEntity;
+    }
+
+    void MatchOverComponent::UpdateRound(uint16_t round)
+    {
+        m_currentRound = round;
+    }
+
+    void MatchOverComponent::DetermineIfMatchEnded(RoundTimeSec roundTime)
+    {
+        
+        const NetworkMatchComponent* netMatchComponent = GetEntity()->FindComponent<NetworkMatchComponent>();
+
+        if (m_currentRound >= netMatchComponent->GetTotalRounds())
+        {
+            auto const currentTime = aznumeric_cast<float>(roundTime);
+            AZ_TracePrintf("MatchOverComponent", "Final round. Awaiting match end in %f\n", currentTime);
+            if (currentTime <= 0)
+            {
+                OnMatchEnd();
```

Comment:
This should okay right now but one thing to worry about is what if `DetermineIfMatchEnded` gets called multiple times by an error from the NetworkMatchComponent.

Reply:
hmm what's the right way to handle that case, disconnect the event handler after this is first triggered?